_source/_posts/2023-09-15-java-spring-boot-microservices.md

```diff
@@ -0,0 +1,722 @@
+
+You can see the access token's contents by copying/pasting it into [jwt.io](https://jwt.io). You can also access the car service directly using it.
+
+```shell
+TOKEN=<access-token>
+http :8090/cars Authorization:"Bearer $TOKEN"
+```
+
+{% img blog/spring-boot-microservices/car-service-with-token.png alt:"Car Service with HTTPie and access token" %}{: .center-image }
+
+Pretty cool, eh? 😎
+
+## My Developer Story with Spring Boot and Spring Cloud
+
+A few years ago, I created an [example similar to this one](http://developer.okta.com/blog/2019/05/22/java-microservices-spring-boot-spring-cloud) with Spring Boot 2.2. It used Feign for remote connectivity, Zuul for routing, Hystrix for failover, and Spring Security for OAuth. The September 2023 version of Spring Cloud has [Spring Cloud OpenFeign](https://spring.io/projects/spring-cloud-openfeign) for remote connectivity, [Spring Cloud Gateway](https://spring.io/projects/spring-cloud-gateway) for routing, and [Resilience4j](https://resilience4j.readme.io/) for fault tolerance.
+
+Okta also now has an [Okta Spring Boot starter](https://github.com/okta/okta-spring-boot). I didn't use it in my first experiment, but I'm a big fan of it after the last few years! It dramatically simplifies configuration and makes securing your apps with OAuth 2.0 and OIDC easy. It's a thin wrapper around Spring Security's resource server, OAuth client, and OIDC features. Not only that, but it works with Okta Workforce Identity, Okta Customer Identity (aka Auth0), and even Keycloak.
+
+I created all of these applications using [start.spring.io](https://start.spring.io)'s REST API and [HTTPie](https://httpie.org).
+
+```shell
+https start.spring.io/starter.zip bootVersion==3.1.4 \
+  artifactId==discovery-service name==eureka-service \
+  dependencies==cloud-eureka-server baseDir==discovery-service | tar -xzvf -
+
+https start.spring.io/starter.zip bootVersion==3.1.4 \
+  artifactId==car-service name==car-service baseDir==car-service \
+  dependencies==actuator,cloud-eureka,data-jpa,data-rest,postgresql,web,validation,devtools,docker-compose,okta | tar -xzvf -
+
+https start.spring.io/starter.zip bootVersion==3.1.4 \
+  artifactId==api-gateway name==api-gateway baseDir==api-gateway \
+  dependencies==cloud-eureka,cloud-feign,data-rest,web,okta | tar -xzvf -
+```` 
+
+You might notice the `api-gateway` project doesn't have `cloud-gateway` as a dependency. That's because I started without it and didn't add it until I needed to proxy requests by path.
+
+After creating these three projects, I ran `chmod +x gradlew` in each directory to make the Gradle wrapper executable.
+
+## Service Discovery with Netflix Eureka
+
+The `discovery-service` is configured the same as you would most Eureka servers. It has an `@EnableEurekaServer` annotation on its main class and properties that set its port and turn off discovery.
+
+```properties
+server.port=8761
+eureka.client.register-with-eureka=false
```

Comment:
In workshop there is also `eureka.client.fetch-registry=false`

Reply:
Good catch!